server/src/main/java/io/druid/server/coordinator/cost/SegmentsCostCache.java

```diff
@@ -0,0 +1,349 @@
+      }
+      cost += bucket.cost(segment);
+    }
+
+    return cost;
+  }
+
+  public static Builder builder()
+  {
+    return new Builder();
+  }
+
+  public static class Builder
+  {
+    private NavigableMap<Interval, Bucket.Builder> buckets = new TreeMap<>(Comparators.intervalsByStartThenEnd());
+
+    public Builder addSegment(DataSegment segment)
+    {
+      Bucket.Builder builder = buckets.computeIfAbsent(toBucketInterval(segment), Bucket::builder);
+      builder.addSegment(segment);
+      return this;
+    }
+
+    public Builder removeSegement(DataSegment segment)
+    {
+      Interval interval = toBucketInterval(segment);
+      buckets.computeIfPresent(
+          interval,
+          (i, builder) -> builder.removeSegment(segment).isEmpty() ? null : builder
+      );
+      return this;
+    }
+
+    public boolean isEmpty()
+    {
+      return buckets.isEmpty();
+    }
+
+    public SegmentsCostCache build()
+    {
+      return new SegmentsCostCache(
+          buckets
+              .entrySet()
+              .stream()
+              .map(entry -> entry.getValue().build())
+              .collect(Collectors.toCollection(ArrayList::new))
+      );
+    }
+
+    private Interval toBucketInterval(DataSegment segment)
+    {
+      long start = segment.getInterval().getStartMillis() - (segment.getInterval().getStartMillis() % BUCKET_INTERVAL);
+      return new Interval(start, start + BUCKET_INTERVAL);
+    }
+  }
+
+  static class Bucket
+  {
+    private final Interval interval;
+    private final ArrayList<DataSegment> sortedSegments;
+    private final double[] leftSum;
+    private final double[] rightSum;
+
+    public Bucket(Interval interval, ArrayList<DataSegment> sortedSegments, double[] leftSum, double[] rightSum)
+    {
+      this.interval = Preconditions.checkNotNull(interval, "interval");
+      this.sortedSegments = Preconditions.checkNotNull(sortedSegments, "sortedSegments");
+      this.leftSum = Preconditions.checkNotNull(leftSum, "leftSum");
+      this.rightSum = Preconditions.checkNotNull(rightSum, "rightSum");
+      Preconditions.checkArgument(sortedSegments.size() == leftSum.length && sortedSegments.size() == rightSum.length);
+      Preconditions.checkArgument(SEGMENT_ORDERING.isOrdered(sortedSegments));
+    }
+
+    public Interval getInterval()
+    {
+      return interval;
+    }
+
+    public boolean inCalculationInterval(DataSegment dataSegment)
+    {
+      return (dataSegment.getInterval().getStartMillis() - interval.getEndMillis() < LIFE_THRESHOLD)
+             && (interval.getStartMillis() - dataSegment.getInterval().getEndMillis() < LIFE_THRESHOLD);
+    }
+
+    public double cost(DataSegment dataSegment)
+    {
+      double cost = 0.0;
+
+      double t0 = convertStart(dataSegment, interval);
+      double t1 = convertEnd(dataSegment, interval);
+
+      // avoid calculation for segments outside of LIFE_THRESHOLD
+      if (!inCalculationInterval(dataSegment)) {
+        return cost;
```

Comment:
"This situation is already excluded on the level where Bucket.cost() is called, so throwing ISE seems more reasonable to me" - prev comment